fe/topology.py

```diff
@@ -15,76 +15,131 @@
+            host_lambda_mult = jnp.zeros(len(host_idxs), dtype=np.int32)
+            host_lambda_offset = jnp.ones(len(host_idxs), dtype=np.int32)
         else:
-            potential.set_idxs(potential.get_idxs() + self.num_host_atoms)
-            return params, potential
+            # (ytz): this extra jank is to work around jnp.concatenate not supporting empty lists.
+            host_params = np.array([], dtype=guest_params.dtype).reshape((-1, guest_params.shape[1]))
+            host_idxs = np.array([], dtype=guest_idxs.dtype).reshape((-1, guest_idxs.shape[1]))
+            host_lambda_mult = []
+            host_lambda_offset = []
+
+        combined_params = jnp.concatenate([host_params, guest_params])
+        combined_idxs = np.concatenate([host_idxs, guest_idxs])
+        combined_lambda_mult = np.concatenate([host_lambda_mult, guest_lambda_mult]).astype(np.int32)
+        combined_lambda_offset = np.concatenate([host_lambda_offset, guest_lambda_offset]).astype(np.int32)
+
+        ctor = type(guest_potential)
+
+        return combined_params, ctor(combined_idxs, combined_lambda_mult, combined_lambda_offset)
 
     def parameterize_harmonic_bond(self, ff_params):
-        return self._parameterize_bonded_term(ff_params, self.guest_topology.parameterize_harmonic_bond)
+        guest_params, guest_potential = self.guest_topology.parameterize_harmonic_bond(ff_params)
+        return self._parameterize_bonded_term(guest_params, guest_potential, self.host_harmonic_bond)
 
     def parameterize_harmonic_angle(self, ff_params):
-        return self._parameterize_bonded_term(ff_params, self.guest_topology.parameterize_harmonic_angle)
-
-    def parameterize_proper_torsion(self, ff_params):
-        return self._parameterize_bonded_term(ff_params, self.guest_topology.parameterize_proper_torsion)
+        guest_params, guest_potential = self.guest_topology.parameterize_harmonic_angle(ff_params)
+        return self._parameterize_bonded_term(guest_params, guest_potential, self.host_harmonic_angle)
 
-    def parameterize_improper_torsion(self, ff_params):
-        return self._parameterize_bonded_term(ff_params, self.guest_topology.parameterize_improper_torsion)
+    def parameterize_periodic_torsion(self, proper_params, improper_params):
+        guest_params, guest_potential = self.guest_topology.parameterize_periodic_torsion(proper_params, improper_params)
+        return self._parameterize_bonded_term(guest_params, guest_potential, self.host_periodic_torsion)
 
     def parameterize_nonbonded(self, ff_q_params, ff_lj_params):
-        # this needs to take care of the case when there's parameter interpolation.
         num_guest_atoms = self.guest_topology.get_num_atoms()
         guest_qlj, guest_p = self.guest_topology.parameterize_nonbonded(ff_q_params, ff_lj_params)
-        
+
+        if isinstance(guest_p, potentials.InterpolatedPotential):
+            assert guest_qlj.shape[0] == num_guest_atoms*2
```

Comment:
Possibly suggests a slight refactoring here, to reduce amount of second-guessing whether `guest_p` is parameter-interpolated or not

Reply:
cleaned this logic part up a bit

Reply:
Nice! Marking resolved